src/components/MultiHopTrade/components/TradeConfirm/hooks/useActiveTradeAllowance.tsx

```diff
@@ -0,0 +1,121 @@
+import { useSignPermit2 } from '../../MultiHopTradeConfirm/hooks/useSignPermit2'
+import { useCurrentHopIndex } from './useCurrentHopIndex'
+
+type UseSignAllowanceApprovalProps = {
+  tradeQuoteStep: TradeQuoteStep
+  isExactAllowance: boolean | undefined
+  activeTradeId: string
+}
+
+export const useActiveTradeAllowance = ({
+  tradeQuoteStep,
+  isExactAllowance,
+  activeTradeId,
+}: UseSignAllowanceApprovalProps) => {
+  // TODO: confirm this is actually needed in the new flow
+  // DO NOT REMOVE ME. Fetches and upserts permit2 quotes at pre-permit2-signing time
+  useGetTradeQuotes()
+
+  const currentHopIndex = useCurrentHopIndex()
+
+  const hopExecutionMetadataFilter = useMemo(() => {
+    return {
+      tradeId: activeTradeId,
+      hopIndex: currentHopIndex,
+    }
+  }, [activeTradeId, currentHopIndex])
+
+  const {
+    state: hopExecutionState,
+    allowanceApproval,
+    allowanceReset,
+    permit2,
+  } = useAppSelector(state => selectHopExecutionMetadata(state, hopExecutionMetadataFilter))
+
+  const { signPermit2 } = useSignPermit2(tradeQuoteStep, currentHopIndex, activeTradeId)
+
+  const isAllowanceApprovalEnabled = useMemo(() => {
+    return (
+      Boolean(allowanceApproval.isRequired) &&
+      [
+        HopExecutionState.Pending,
+        HopExecutionState.AwaitingAllowanceReset,
+        HopExecutionState.AwaitingAllowanceApproval,
+      ].includes(hopExecutionState)
+    )
+  }, [allowanceApproval.isRequired, hopExecutionState])
+
+  const {
+    approveMutation,
+    approvalNetworkFeeCryptoBaseUnit,
+    isLoading: isAllowanceApprovalLoading,
+  } = useAllowanceApproval(
+    tradeQuoteStep,
+    currentHopIndex,
+    // Permit2 should always have unlimited allowance
+    isExactAllowance && !permit2.isRequired ? AllowanceType.Exact : AllowanceType.Unlimited,
+    isAllowanceApprovalEnabled,
+    activeTradeId,
+    allowanceApproval.isInitiallyRequired,
+  )
+
+  const isAllowanceResetEnabled = useMemo(() => {
+    return (
+      Boolean(allowanceReset.isRequired) &&
+      [HopExecutionState.Pending, HopExecutionState.AwaitingAllowanceReset].includes(
+        hopExecutionState,
+      )
+    )
+  }, [allowanceReset.isRequired, hopExecutionState])
+
+  const {
+    allowanceResetMutation,
+    allowanceResetNetworkFeeCryptoBaseUnit,
+    isLoading: isAllowanceResetLoading,
+  } = useAllowanceReset(
+    tradeQuoteStep,
+    currentHopIndex,
+    AllowanceType.Reset,
+    isAllowanceResetEnabled,
+    activeTradeId,
+    allowanceReset.isInitiallyRequired,
+  )
+
+  const handleSignAllowanceApproval = useCallback(async () => {
+    try {
+      await approveMutation.mutateAsync()
+    } catch (error) {
+      console.error(error)
+    }
+  }, [approveMutation])
+
+  const handleSignAllowanceReset = useCallback(async () => {
+    try {
+      await allowanceResetMutation.mutateAsync()
+    } catch (error) {
+      console.error(error)
+    }
+  }, [allowanceResetMutation])
```

Comment:
nitpick: since we're not really concerned about handling errors, and all we care about is firing the mutation, we can use `mutate()` instead of `mutateAsync()`, error-handling will be handled internally by react-query

```suggestion
  const handleSignAllowanceApproval = approveMutation.mutate
```

Reply:
Yeah that's fair! https://github.com/shapeshift/web/commit/20c18f0b55d5c38c8aff1f9415eddde956816bf8